JSTests/stress/bounds-check-zlib.js

```diff
@@ -0,0 +1,90 @@
+//@ runDefault("--useConcurrentJIT=0", "--useDollarVM=1", "--jitAllowList=provableRedundantBitshift,provableRedundantEasy")
+
+// jsc JSTests/stress/bounds-check-zlib.js --useConcurrentJIT=0 --useDollarVM=1 --jitAllowList=provableRedundantBitshift,provableRedundantEasy --dumpDisassembly=0
+// Based on:
+// jsc -e "testList = ['octane-zlib'];" cli.js --useConcurrentJIT=0 --jitAllowList=a6#A8KgFt --dumpDisassembly=1 --forceICFailure=1
+
+let arr = Array(1000).fill(0)
+
+let tests = []
+
+function provableRedundantEasy(b, d) {
+    "use strict";
+
+    b = b | 0
+    d = d | 0
+
+    if (arr.length != 1000)
+        return;
+
+    let sum = 0
+
+    if (d > 499)
+        return sum
+
+    // There should be no checks.
+    sum += arr[d + 500] | 0
+    sum += arr[d + 400] | 0
+
+    return sum
+}
+noInline(provableRedundantEasy)
+// TODO: these checks should be removed.
+tests.push({ fn: provableRedundantEasy, before: 2, after: 2 })
+
+function provableRedundantBitshift(b, d) {
+    "use strict";
+
+    b = b | 0
+    d = d | 0
+
+    if (arr.length != 1000)
+        return;
+
+    let sum = 0
+
+    if ((d >>> 2) > 10)
+        return sum
+
+    sum += arr[d + 500] | 0
+    sum += arr[d + 400] | 0
+    sum += arr[(b + 500) >>> 2] | 0
+    sum += arr[(b + 500) >>> 3] | 0
+    sum += arr[(b + 500) >>> 4] | 0
+    sum += arr[(b + 400) >>> 2] | 0
+    sum += arr[(b + 300) >>> 2] | 0
+    sum += arr[(b + 200) >>> 2] | 0
+    sum += arr[(b + 100) >>> 2] | 0
+    sum += arr[(b + 50) >>> 2] | 0
+
+    sum += arr[(b + 50) >>> 2] | 0
+
+    sum += arr[(d + 500) >>> 2] | 0
+    sum += arr[(d + 500) >>> 3] | 0
+    sum += arr[(d + 500) >>> 4] | 0
+    sum += arr[(d + 400) >>> 2] | 0
+
+    return sum
+}
+noInline(provableRedundantBitshift)
+tests.push({ fn: provableRedundantBitshift, before: 6, after: 6 })
+
+function main() {
+    for (test of tests) {
+        $vm.startDebugRecordingIRO()
+        for (let i = 0; i < 20000000; ++i) {
+            test.fn(i, i % 10)
+        }
+        let result = $vm.stopDebugRecordingIRO()
+        let before = result & 0xFFFFFFFFn;
+        let after = (result >> 32n) & 0xFFFFFFFFn;
+
+        if (before != test.before)
+            throw "Test " + test.fn + " failed: expected " + test.before + " checks before IRO, found " + before
+        if (after != test.after)
+            throw "Test " + test.fn + " failed: expected " + test.after + " checks after IRO, found " + after
```

Comment:
I think you'll want to refactor this test so that if there's not DFG it still passes e.g. on watchOS.

Reply:
I am planning to skip it